### PR TITLE
Add generic diagnostic redactor back to v2

### PR DESCRIPTION
**Pull request by joshdover**
## What does this PR do?

Restores the redaction step from https://github.com/elastic/elastic-agent/pull/566 during diagnostic building to ensure that most secrets in configuration are not included in the diag.

There are situations where the YAML is not always unmarshallable https://github.com/elastic/elastic-agent/issues/1907#issuecomment-1363224642 which limits the ability to redact the agent's `state.yml` diag file. This will need to be fixed by either a smarter generic redaction implementation that can selectively unmarshall/redact parts of the file or by redacting the state.yml file before it's serialized in `coordinator.go`.

In the meantime, I have made redaction best-effort and will allow any file that cannot be successfully unmarshalled to be written to the diag as-is with secrets in place. A warning log is written to stderr with details about which files were not redacted when this happens.

~~This also fixes a bug where diagnostic files that provided by components were written to the diag bundle without the file extension.~~ Got fixed in https://github.com/elastic/elastic-agent/pull/2047

## Why is it important?

We do not want users to include any secrets in the diagnostics that they end to us. We have had this feature since 8.4 and do not want to regress on it

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Relates to https://github.com/elastic/elastic-agent/issues/1907

### CODE REVIEW

**Comment by cmacknz** [comment]
@michalpristas might make sense for you to finish this one off since I believe Josh is off most or all of next week.

**Comment by cmacknz** [comment]
Nobody was able to get to this over the past week. I would not be opposed to merging this without the redaction in state.yml if it comes to it, I'd rather have one unredacted file then many.

**Comment by joshdover** [comment]
> Edit: Tested this locally, not quite ready to be merged as is.

Do you remember what issue you noticed? I may have time to get at least this part in before the final BC

File: internal/pkg/agent/cmd/diagnostics.go [hunk 1]
```diff
@@ -156,6 +163,72 @@ func createZip(fileName string, agentDiag []client.DiagnosticFileResult, unitDia
...
+			// Best effort, output a warning but still include the file
+			fmt.Fprintf(streams.Err, "[warning] Could not redact %s due to unmarshalling error: %s\n", fullFilePath, err)
+		} else {
+			redacted, err := yaml.Marshal(redactMap(unmarshalled))
+			if err != nil {
+				// Best effort, output a warning but still include the file
+				fmt.Fprintf(streams.Err, "[warning] Could not redact %s due to marshalling error: %s\n", fullFilePath, err)
+			}
+	
+			out = &redacted
```
Comment (joshdover) [inline internal/pkg/agent/cmd/diagnostics.go:184]: This needs to be an in else to avoid modifying when there is a marshalling error

**Comment by cmacknz** [comment]
> @cmacknz Do you remember what issue you noticed? I may have time to get at least this part in before the final BC

No, I was actually experimenting with the yaml.v3 package version to see if it fixed the problem unmarshalling the keys. I was seeing sections of the diagnostics unmarshal to empty maps like `{}` but I'm not sure if that happened with the original `yaml.v2`. The problem was obvious and not subtle so if it looks good to you it's likely fine. I tried this quickly and not thoroughly. 

I wanted to try lazily unmarshalling the state.yaml file and just avoid the section that had the error but didn't have enough time.

**Comment by joshdover** [comment]
> I was seeing sections of the diagnostics unmarshal to empty maps like `{}` but I'm not sure if that happened with the original `yaml.v2`. The problem was obvious and not subtle so if it looks good to you it's likely fine.

I haven't been able to reproduce this with v2. This appears to still be working with the caveat of state.yml. I'm going to put this up for review, but not close the underlying bug yet. We will need to spend some effort fully fixing this for an 8.6.x patch release.

File: internal/pkg/agent/cmd/diagnostics.go [hunk 2]
```diff
@@ -156,6 +163,72 @@ func createZip(fileName string, agentDiag []client.DiagnosticFileResult, unitDia
...
 
+func writeRedacted(streams *cli.IOStreams, fullFilePath string, fr client.DiagnosticFileResult, w io.Writer) error {
+	out := &fr.Content
+
+	// Should we support json too?
+	if fr.ContentType == "application/yaml" {
+		unmarshalled := map[string]interface{}{}
+		err := yaml.Unmarshal(fr.Content, &unmarshalled)
+
+		if err != nil {
```
Comment (AndersonQ) [inline internal/pkg/agent/cmd/diagnostics.go:174]: [Go convention]
```suggestion
		err := yaml.Unmarshal(fr.Content, &unmarshalled)
		if err != nil {
```

File: internal/pkg/agent/cmd/diagnostics.go [hunk 3]
```diff
@@ -156,6 +163,72 @@ func createZip(fileName string, agentDiag []client.DiagnosticFileResult, unitDia
...
+			// Best effort, output a warning but still include the file
+			fmt.Fprintf(streams.Err, "[warning] Could not redact %s due to unmarshalling error: %s\n", fullFilePath, err)
```
Comment (AndersonQ) [inline internal/pkg/agent/cmd/diagnostics.go:176]: [Suggestion]
Make it upper case in the hope it'll help it'll draw better people's attention

```suggestion
			fmt.Fprintf(streams.Err, "[WARNING] Could not redact %s due to unmarshalling error: %s\n", fullFilePath, err)
```

**Review by AndersonQ — APPROVED** [review]
It'd be nice to have a test, but due to the time constraint, it's good to go

**Review by michalpristas — APPROVED** [review]

File: internal/pkg/agent/cmd/diagnostics.go [hunk 4]
```diff
@@ -156,6 +163,72 @@ func createZip(fileName string, agentDiag []client.DiagnosticFileResult, unitDia
...
+		} else {
+			redacted, err := yaml.Marshal(redactMap(unmarshalled))
+			if err != nil {
+				// Best effort, output a warning but still include the file
+				fmt.Fprintf(streams.Err, "[warning] Could not redact %s due to marshalling error: %s\n", fullFilePath, err)
```
Comment (joshdover) [inline internal/pkg/agent/cmd/diagnostics.go:181]:
```suggestion
				fmt.Fprintf(streams.Err, "[WARNING] Could not redact %s due to marshalling error: %s\n", fullFilePath, err)
```

**Pull request merged**